addrmgr: Decouple address manager from wire.NetAddress

sefbkn opened the pull request:
This change decouples the address manager's internal handling of network addresses to use a type owned by the address manager.

dnldd commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+			reach:         Ipv4,
+		},
+		{
+			name:          "not routable ipv4 to torv2",
```
Perhaps "not routable" in the test names could be "unroutable".

sefbkn replied:
Right, that's more consistent with the rest of the code. Thanks, updated.

dnldd commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+			reach:         Default,
+		},
+		{
+			name:          "not routable to routable",
```
s / "not routable to routable" / "unroutable ipv4 to routabe ipv6"

sefbkn replied:
Updated

dnldd commented on addrmgr/netaddress.go:
```diff
@@ -0,0 +1,89 @@
+	return NewAddressFromWireNetAddress(na), nil
+}
+
+// NewNetAddressFromString creates an addrmgr net address from the provided
```
s / ... addrmgr / address manager.

sefbkn replied:
Fixed

dnldd commented on addrmgr/netaddress.go:
```diff
@@ -0,0 +1,89 @@
+	return newNetAddr
+}
+
+// NewAddressFromHostPort creates an addrmgr net address from the provided
```
s / ... addrmgr ... / ... address manager ...

sefbkn replied:
Fixed

dnldd commented on server.go:
```diff
@@ -602,14 +603,32 @@ func (sp *serverPeer) relayTxDisabled() bool {
+	return addrs
+}
+
+// addrmgrToWireNetAddress is a helper function that converts an to addrmgr net
```
s / ... addrmgr ... / ... address manager ...

sefbkn replied:
Updated

dnldd reviewed (commented):
Nice, looks good overall.

rstaudt2 commented on addrmgr/addrmanager_test.go:
```diff
@@ -202,7 +208,7 @@ func TestAddAddressByIP(t *testing.T) {
 	amgr.Start()
 	if ka := amgr.GetAddress(); ka == nil {
-		t.Errorf("Address Manager should contain known address")
+		t.Fatalf("Address Manager should contain known address")
```
This (and several other cases below) can be `Fatal` rather than `Fatalf` since no formatting is required.

rstaudt2 reviewed (commented):
Great to see more test coverage being added as part of this decoupling!

sefbkn commented on addrmgr/addrmanager.go:
```diff
@@ -1197,11 +1189,12 @@ func (a *AddrManager) ValidatePeerNa(localAddr, remoteAddr *wire.NetAddress) (bo
+		rand:            rand.New(rand.NewSource(time.Now().UnixNano())),
+		quit:            make(chan struct{}),
+		localAddresses:  make(map[string]*localAddress),
+		triedBucketSize: 256,
```
Agreed that would be much clearer than a hardcoded number in a constructor.

sefbkn replied:
Updated

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+		default:
+			panic("invalid test case: " + host)
+		}
+	}
```
I think it makes sense to keep it scoped closer to the test case, will update.

sefbkn replied:
Updated to define the func on each test.

sefbkn commented:
Thanks for the feedback @dnldd @rstaudt2. I've not had the opportunity to wrap up applying the suggested changes yet, but will be able to shortly.

sefbkn commented on addrmgr/addrmanager.go:
```diff
@@ -49,6 +49,18 @@ type AddrManager struct {
+
+	// triedBucketSize is the maximum number of addresses in each
+	// tried address bucket.
+	triedBucketSize int
```
Updated. Also updated comments throughout this file.

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -271,6 +299,27 @@ func TestAddLocalAddress(t *testing.T) {
+
+		if test.valid && !amgr.HasLocalAddress(&test.address) {
+			t.Errorf("TestAddLocalAddress test #%d failed: expected to have "+
+				"local address", x)
```
Updated

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -294,6 +343,11 @@ func TestAttempt(t *testing.T) {
+	// Attempt an ip not known to the address manager.
+	unknownIP := net.ParseIP("1.2.3.4")
+	unknownNetAddress := wire.NewNetAddressIPPort(unknownIP, 1234, wire.SFNodeNetwork)
+	n.Attempt(unknownNetAddress)
```
Updated. Previously this was just to exercise the path that prevents a nil panic. Now `Attempt` returns an error if the address is not found. Same pattern for `Good` and `Connected`.

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -314,6 +368,12 @@ func TestConnected(t *testing.T) {
+	// connected.
+	unknownIP := net.ParseIP("1.2.3.4")
+	unknownNetAddress := wire.NewNetAddressIPPort(unknownIP, 1234, wire.SFNodeNetwork)
+	n.Connected(unknownNetAddress)
```
Updated to return an error.

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -378,6 +438,66 @@ func TestGood(t *testing.T) {
+	}
+	if _, exists := n.addrNew[0][addrBKey]; !exists {
+		t.Errorf("Expected addrB to exist in internal addrmgr.addrNew bucket")
+	}
```
Updated

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -378,6 +438,66 @@ func TestGood(t *testing.T) {
+		t.Errorf("Expected addrmgr.addrTried bucket to contain one element.")
+	}
+	if NetAddressKey(n.addrTried[0][0].na) != addrBKey {
+		t.Errorf("Expected addrA to exist in addrmgr.addrTried bucket")
```
Fixed

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+
+		valid, reach := addressManager.ValidatePeerNa(localNa, remoteNa)
+		if valid != test.valid {
+			t.Errorf("TestValidatePeerNa (%s): unexpected return value for "+
```
Updated

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+		}
+		if test.expected != nil && (result.Services != test.services ||
+			!bytes.Equal(result.IP, test.expected.IP) ||
+			result.Port != test.expected.Port) {
```
Makes sense, updated.

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +678,328 @@ func TestCorruptPeersFile(t *testing.T) {
+			shouldError: false,
+			expected: wire.NewNetAddressIPPort(
+				net.ParseIP("fd87:d87e:eb43:744:208d:5408:63a4:ac4f"), 8333,
+				wire.SFNodeNetwork),
```
Updated

sefbkn commented:
Since there are changes in this PR that were not in the initial commits, I want to call out some high-level differences although most of the functional changes are the same plus a little more to fully decouple the addrmgr module. The aim was to help make reviewing it easier since a lot of changes were squashed into a single commit prevoiously

- Added more detailed comments in the address manager
- Remove some unnecessary tests and move others to netaddress_test.go for methods that moved to NetAddress
- Split out the changes into more commits rather than squashing too many changes in one larger commit at the end
- Completely decouple the address manager module from wire

dnldd commented on addrmgr/addrmanager.go:
```diff
@@ -900,86 +999,95 @@ func (a *AddrManager) Good(addr *wire.NetAddress) {
 	// record one of the buckets in question and call it the `first'
 	addrKey := NetAddressKey(addr)
-	oldBucket := -1
+	availableNewBucketIndex := -1
```
`availableNewBucketIndex` is a confusing variable name for me because it does not refer to `addrNew` explicitly, calling it a new bucket instead. I do not have a good variable name at the moment. Something along the lines of `addrNewOpenSlot` or `addrNewAvailableIndex`   would be better I think.

sefbkn replied:
Updated

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +724,325 @@ func TestCorruptPeersFile(t *testing.T) {
+
+// TestHostToNetAddress ensures that HostToNetAddress behaves as expected
+// given valid and invalid host name arguments.
+func TestHostToNetAddress(t *testing.T) {
```
Rather than adding a test here, `HostToNetAddress` can be moved out of the address manager and moved to a private method on the main module along with these test cases. This also allows removing the need to pass in a `lookupFunc` to `New`.

sefbkn replied:
This does not seem necessary for this PR, deferring. Though if anyone has preference to see this change made here, adding it would not be an issue.

sefbkn commented on addrmgr/network.go:
```diff
@@ -116,18 +116,19 @@ func isOnionCatTor(netIP net.IP) bool {
 
 const (
-	LocalAddress NetworkAddress = iota
+	LocalAddress NetAddressType = iota
```
`LocalAddress` can be removed. Since it's only used in `getNetwork` which is called from a single place, the call to `isLocal` can be moved up the stack. In it's place for a zero value, we can use something like `UnknownAddressType` to represent an address whose type cannot be determined.

sefbkn replied:
This can be deferred since it's not strictly necessary for the decoupling and the field is not used at this point anyways.

dnldd commented on addrmgr/addrmanager.go:
```diff
@@ -900,86 +999,95 @@ func (a *AddrManager) Good(addr *wire.NetAddress) {
-		// What? wasn't in a bucket after all.... Panic?
-		return
+	if availableNewBucketIndex == -1 {
+		str := fmt.Sprintf("address %s not found in new bucket",
```
s / `address %s not found in new bucket` / `%s is not marked as a new address`

dnldd commented on server.go:
```diff
@@ -686,7 +686,10 @@ func (sp *serverPeer) OnVersion(_ *peer.Peer, msg *wire.MsgVersion) {
-		addrManager.SetServices(remoteAddr, msg.Services)
+		err := addrManager.SetServices(remoteAddr, msg.Services)
+		if err != nil {
+			srvrLog.Debugf("Setting services for address failed: %v", err)
```
What's the reason for preferring `Debugf` here instead of `Errorf` here and the instances below? Should we not be returning on these errors?

sefbkn replied:
`Errorf` could be more appropriate since the address should be in the address manager at this point, updated.

As far as returning and disconnecting, it wasn't returning prior to this change if the address was not in the address manager so I left it to follow the same path as a part of this refactor.

sefbkn replied:
It appears that permanent peers were/are not added to the address manager so I reverted this back to `Debugf`.  There might be some other category of addresses also not added to the address manager but I'm not entirely sure. Although this does need to be accounted for so this branch is not hit, my intention for this PR was non-functional changes.

dnldd replied:
Noted, and yes I think the peer should be disconnected from if `SetServices` or `Good` fails.

dnldd commented on addrmgr/addrmanager_test.go:
```diff
@@ -108,80 +108,66 @@ func lookupFunc(host string) ([]net.IP, error) {
-	dir, err := ioutil.TempDir("", "testaddressbyip")
+// TestStartStop tests the behavior of the address manager when it is started
+// and stopped.
+func TestStartStop(t *testing.T) {
```
This is probably not the PR to this in but since you are working on the address manager you should consider converting it to leverage contexts and ditch start/stop process.

rstaudt2 reviewed (commented):
Updates look good to me!  I also ran `./run_tests.sh` on each commit and tested various scenarios manually and did not see any issues.

sefbkn commented on addrmgr/addrmanager.go:
```diff
@@ -1040,6 +1153,7 @@ func (a *AddrManager) LocalAddresses() []LocalAddr {
 		la := LocalAddr{
 			Address: addr.na.IP.String(),
 			Port:    addr.na.Port,
+			Score:   0,
```
Updated

davecgh commented on addrmgr/addrmanager.go:
```diff
@@ -879,18 +974,22 @@ func (a *AddrManager) Connected(addr *wire.NetAddress) {
-// manager it will be ignored.
-func (a *AddrManager) Good(addr *wire.NetAddress) {
+// Good marks the given address as good.  This should be called after a
+// successful outbound connection to and version exchange with a peer.  If the
```
```suggestion
// successful outbound connection and version exchange with a peer.  If the
```

davecgh commented on addrmgr/addrmanager.go:
```diff
@@ -900,86 +999,94 @@ func (a *AddrManager) Good(addr *wire.NetAddress) {
-	// We don't touch a.nTried here since the number of tried stays the same
-	// but we decremented a.nNew above, raise it again since we're putting
-	// something back.
+	// The total number of tried addresses a.nTried is not modified here since
```
```suggestion
	// The total number of tried addresses is not modified here since
```

It's best practice to avoid referencing variable names in comments since they often change over time and the comments don't get updated.  Instead, they should describe overall intent.

sefbkn replied:
That makes sense, thanks for the tip. Updated comments. Some of the tests were also referencing variable names and those were updated as well.

davecgh commented on addrmgr/addrmanager.go:
```diff
@@ -900,86 +999,94 @@ func (a *AddrManager) Good(addr *wire.NetAddress) {
-	// something back.
+	// The total number of tried addresses a.nTried is not modified here since
+	// the number of tried addresses stays the same.  However, since the total
+	// number of new addresses a.nNew was decremented above, increment it now
```
Same here.
```suggestion
	// number of new addresses was decremented above, increment it now
```

davecgh commented on addrmgr/addrmanager.go:
```diff
@@ -991,13 +1098,16 @@ func (a *AddrManager) SetServices(addr *wire.NetAddress, services wire.ServiceFl
 func (a *AddrManager) AddLocalAddress(na *wire.NetAddress, priority AddressPriority) error {
 	if !IsRoutable(na) {
-		return fmt.Errorf("address %s is not routable", na.IP)
+		return fmt.Errorf("address %s is not routable", ipString(na))
```
Perhaps it's coming up in a future commit, but I really think the new type that is local to the address manager should end up with a `fmt.Stringer` on it so the address can just be printed like normal.

sefbkn replied:
The new type satisfies the `fmt.Stringer` interface.

dnldd replied:
if that's the case then `ipString` helper would not be needed to stringify `na`, just providing `na` should be adequate.

sefbkn replied:
Right, in the later commits that take advantage of the stringer interface the `ipString` call is dropped.  Two things to note here are that `ipString` converts TORv2 addresses to a friendly .onion hostname rather than printing as an IPv6 address, and that in the later commit `String` includes both the host and port rather than just the IP address.  My thinking here is that:

1. If the address is a tor address, it should print the ".onion" formatting rather than the IPv6 formatting (this commit)
2. The net address is uniquely identifiable by host:port so it should be printed with the port rather than just the host (handled by `String()` on the net address type in a later commit)

Plus, there is some overlap with the naming & behavior (of host:port) so the netaddress type can satisfy the net.Addr interface by adding a `Network` method when needed and switching that value based on the address type.

davecgh reviewed (commented):
Thanks for taking this on!  I'll be reviewing this in stages.  First round of comments inline.

sefbkn commented on addrmgr/addrmanager.go:
```diff
@@ -156,8 +221,8 @@ const (
 	// getAddrMax is the most addresses that we will send in response
-	// to a getAddr (in practice the most addresses we will return from a
-	// call to AddressCache()).
+	// to a getAddr message. Effectively, this defines the most addresses
```
Right, updated.

sefbkn replied:
I changed the wording on this to be more generic since the actual limit for an addr protocol message differs from this constant, and the number of addresses pushed to the peer are limited to the appropriate value elsewhere (for the current protocol version).

dnldd commented on addrmgr/error.go:
```diff
@@ -0,0 +1,45 @@
+// These constants are used to identify a specific RuleError.
+const (
+	// ErrAddressNotFound indicates that an operation in the address manager
+	// failed due to encountering an unknown address.
```
I don't think `ErrAddressNotFound` is a failure related to encountered an unknown address as the comment states, it's an address lookup failure.

sefbkn replied:
Updated

dnldd reviewed (approved):
Minor comment related issues to fix but overall I think this is good to go.

davecgh commented on addrmgr/netaddress.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+// ServiceFlag identifies services supported by a Decred peer.
+type ServiceFlag uint64
```
Why is this being duplicated here?  The supported services by a peer are not something defined by the address manager, rather they are part of the wire protocol.

sefbkn replied:
This was done to break the cyclical dependency between wire and addrmgr.  Otherwise, addrmgr would need to import wire for `wire.ServiceFlag` and wire would need to import addrmgr for `addrmgr.NetAddress`

davecgh replied:
Why would `wire` depend on the new type though?  I would expect `wire` to continue to have its own protocol structure for sending information across the wire that would be converted as needed when going from the address manager's internal representation to wire representation and vice versa.

Please correct me if I'm wrong, but it seems to me that the primary aim of this PR is to "Decouple address manager from `wire.NetAddress", per the title, but having `wire` use the address manager's new `NetAddress` type means it just reverses the coupling.  Specifically, it would mean that `wire` then is coupled to `addrmgr.NetAddress` instead of `addrmgr` being coupled to `wire.NetAddress`, but the coupling would still be present.

I would expect the code in server such as `func (sp *serverPeer) pushAddrMsg` to accept the new `addrmgr.NetAddress` type and convert them in the loop that filters them.  Similarly, I would expect the code in `func (sp *serverPeer) OnVersion` to convert the `wire.NetAddress` to an `addrmgr.NetAddress`, etc.   That would ensure they are actually decoupled.

davecgh replied:
Following up, I should also mention that wire *can't* change the type of the `addr` message without also bumping the major version, which we really, really, don't want to do because it would snowball into having to bump the major of every single API that uses **any** `wire` type, such as `wire.MsgBlock` and `wire.MsgTx` which is basically _everything_.  The fallout from that would be monumental.

Another thing is that the old `addr` message in `wire` will still need to be supported for a time, and it seems really odd to me that it would be defined in terms of something that supports things that the message itself doesn't.

sefbkn replied:
> Why would `wire` depend on the new type though? I would expect `wire` to continue to have its own protocol structure for sending information across the wire that would be converted as needed when going from the address manager's internal representation to wire representation and vice versa.

My intention here was to have `addrv2` depend on `addrmgr.NetAddress` so that the `Type` field could be included for each address in the message.  With the provided feedback in mind, it does seem like a new address type in `wire` could be introduce that is strictly used for serialization/deserialization for `addrv2`.  This way `addrmgr` still depends on wire, but only for `wire.ServiceFlag` and `wire` remains non-dependent on `addrmgr`.  Either way `addrmgr.ServiceFlag` can be removed in this PR as-is.

> I would expect the code in server such as `func (sp *serverPeer) pushAddrMsg` to accept the new `addrmgr.NetAddress` type and convert them in the loop that filters them. Similarly, I would expect the code in `func (sp *serverPeer) OnVersion` to convert the `wire.NetAddress` to an `addrmgr.NetAddress`, etc. That would ensure they are actually decoupled.

Yes, this is the way it is implemented now though the intention was to have `addrv2` depend on `addrmgr.NetAddress` rather than adding a new `wire` type for `addrv2`.  Additionally, my intent was to leave the `addr` implementation alone and dependent on `wire.NetAddress` with the mapping as described.

> Following up, I should also mention that wire can't change the type of the addr message without also bumping the major module version of wire, which we really, really, don't want to do because it would snowball into having to bump the major of every single API that uses any wire type, such as wire.MsgBlock and wire.MsgTx which is basically everything. The fallout from that would be monumental.

Right, the `wire` module has not been modified in this PR and the intent was to leave `addr` dependent on `wire.NetAddress`.  My takeaway here is that adding a new type would avoid the need to bump the `wire` module version for modules that do NOT consume the new type, is this a correct assumption to make?

sefbkn replied:
The addrmgr module has been modified to re-couple it to the wire module while also removing the dependency on wire.NetAddress, rather than inverting the dependency.  The mapping is still handled in server.go as it was before to convert the types across modules.

davecgh commented on addrmgr/addrmanager_test.go:
```diff
@@ -15,95 +15,11 @@ import (
 var someIP = "173.194.115.66"
 
-// addNaTests
-func addNaTests() {
```
Just want to add that I'm glad to see this moved into a properly scoped test.

davecgh commented on addrmgr/netaddress_test.go:
```diff
@@ -0,0 +1,121 @@
+	netAddrClone := netAddr.Clone()
+
+	if netAddr == netAddrClone {
+		t.Error("expected new network address reference")
```
Probably should be `t.Fatal` since the code after this assumes it passed.  In general, `t.Error`/`t.Errorf` are typically used in table-driven tests along with `continue` so the next independent test is executed, but whenever it's a straight-line test like this, the following code typically depends on the result of the previous and thus `t.Fatal`/`t.Fatalf` are more appropriate.

davecgh commented on addrmgr/netaddress_test.go:
```diff
@@ -0,0 +1,121 @@
+		t.Error("expected new network address reference")
+	}
+	if !reflect.DeepEqual(netAddr, netAddrClone) {
+		t.Errorf("unxpected clone result -- got %v, want %v",
```
Also `t.Fatalf`.

davecgh commented on addrmgr/netaddress_test.go:
```diff
@@ -0,0 +1,121 @@
+	netAddr.AddService(sfNodeNetwork)
+
+	if netAddr.Services != sfNodeNetwork {
+		t.Errorf("expected service flag to be set -- got %x, want %x",
```
`t.Fatalf`

davecgh commented on addrmgr/network.go:
```diff
@@ -251,7 +249,7 @@ func isValid(netIP net.IP) bool {
 // IsRoutable returns whether or not the passed address is routable over
 // the public internet.  This is true as long as the address is valid and is not
 // in any reserved ranges.
-func IsRoutable(netIP net.IP) bool {
```
I think this should remain exported.  It is generally useful and, iirc, used by at least wallet.

sefbkn replied:
Updated to re-expose function.

davecgh reviewed (changes requested):
I've finished reviewing the remaining pieces.  Very nice job overall!  I've called out a few minor things inline, but overall the changes make sense and the increased test coverage is a welcome change.  I also like how various methods like `GroupKey` are now defined on the new address type itself instead of being standalone funcs.

The only remaining thing that I think really isn't correct is the addition of the `ServiceFlag` type as called out inline.

Once these last bits are resolved, I'll do some live testing and barring any issues found there, this will be good to go from my viewpoint.  Thanks for your work on this and to everyone who helped review.

sefbkn commented on addrmgr/addrmanager_test.go:
```diff
@@ -558,3 +748,325 @@ func TestCorruptPeersFile(t *testing.T) {
+		remoteAddress string
+		valid         bool
+		reach         int
+	}{
```
Updated to use more compact syntax here and in other modified test files.

sefbkn commented on addrmgr/network.go:
```diff
@@ -118,18 +118,19 @@ func isOnionCatTor(netIP net.IP) bool {
 
 // getNetwork returns the network address type of the provided network address.
-func getNetwork(netIP net.IP) NetworkAddress {
+func getNetwork(netIP net.IP) NetAddressType {
```
Updated.  There are changes in a later pull request to calculate and store this value on the addrmgr.NetAddress type as on construction of the addrmgr.NetAddress, but I'm holding off on back porting that for now.  Let me know if it would be desirable for that to be pulled into this commit or PR.  Otherwise, I'll just stick with renaming the function for now as suggested.

davecgh reviewed (approved):
I've reviewed all the updates.  This looks good to go!
